Sporadic Failure in HDFSIntegrationTest.testEndToEndWithFencing

**Problem description**
The exception causing the test failure is:
```
java.util.concurrent.TimeoutException
    at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
    at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
    at io.pravega.segmentstore.server.store.StreamSegmentStoreTestBase.testEndToEndWithFencing(StreamSegmentStoreTestBase.java:296)
```
Following the flow of the test, it can be seen that everything goes as expected until it executes `waitForSegmentsInStorage`, which is the point where the timeout exception arises.

When the test gets into this method, we start observing `BadOffsetException` messages related with a particular transaction segment:

```
2018-09-18 06:01:51,980 INFO  [storage-io-2] writer.SegmentAggregator (SegmentAggregator.java:setState(1549)) - StorageWriter[3-556]: State changed from Writing to ReconciliationNeeded.
2018-09-18 06:01:51,980 ERROR [storage-io-2] writer.StorageWriter (StorageWriter.java:logError(457)) - StorageWriter[3]: Iteration[2].Error.
io.pravega.segmentstore.contracts.BadOffsetException: [Segment 'Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0'] Bad Offset. Expected 1700, given 0.
    at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorage.write(HDFSStorage.java:360)
    at io.pravega.segmentstore.storage.rolling.RollingStorage.write(RollingStorage.java:322)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$write$2(AsyncStorageWrapper.java:100)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$runAsync$15(AsyncStorageWrapper.java:206)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.execute(AsyncStorageWrapper.java:221)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$supplyAsync$13(AsyncStorageWrapper.java:181)
```
As well as `IllegalArgumentException` in storage writer related to offsets:
```
2018-09-18 06:01:52,997 INFO  [IPC Server handler 6 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true    ugi=root (auth:SIMPLE)    ip=/127.0.0.1    cmd=listStatus    src=/    dst=null    perm=null    proto=rpc
2018-09-18 06:01:53,022 ERROR [core-4] writer.StorageWriter (StorageWriter.java:logError(457)) - StorageWriter[3]: Iteration[3].Error.
java.lang.IllegalArgumentException: Offset 0 is beyond the last offset 0 of the segment.
    at com.google.common.base.Preconditions.checkArgument(Preconditions.java:323)
    at io.pravega.segmentstore.storage.rolling.RollingStorage.read(RollingStorage.java:151)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$read$8(AsyncStorageWrapper.java:135)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.execute(AsyncStorageWrapper.java:221)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$supplyAsync$13(AsyncStorageWrapper.java:181)
```
The test continues trying to read the `SegmentProperties` of segments in HDFS until the `120s` timeout for `waitForSegmentsInStorage` is exhausted and the test fails.

**Problem location**
`HDFSIntegrationTest.testEndToEndWithFencing`

**Suggestions for an improvement**
Make the test to pass consistently.

Here is some relevant output to this discussion, which I'm not seeing in previous comments:

```
# grep 'Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_' testEndToEndWithFencing.txt
2018-09-18 06:01:49,057 INFO  [IPC Server handler 8 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=create	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_0	dst=null	perm=root:supergroup:rw-------	proto=rpc
2018-09-18 06:01:49,071 INFO  [IPC Server handler 3 on 51943] hdfs.StateChange (FSDirWriteFileOp.java:completeFile(605)) - DIR* completeFile: /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_0 is closed by DFSClient_NONMAPREDUCE_2144198447_7088
2018-09-18 06:01:49,698 INFO  [IPC Server handler 3 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=rename	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_0	dst=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2	perm=root:supergroup:rw-------	proto=rpc
2018-09-18 06:01:49,701 INFO  [IPC Server handler 7 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=append	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2	dst=null	perm=null	proto=rpc
2018-09-18 06:01:49,797 INFO  [IPC Server handler 0 on 51943] hdfs.StateChange (FSDirWriteFileOp.java:logAllocatedBlock(725)) - BLOCK* allocate blk_1073741945_1222, replicas=127.0.0.1:35859 for /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2
2018-09-18 06:01:49,853 INFO  [IPC Server handler 0 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=rename	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2	dst=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3	perm=root:supergroup:rw-------	proto=rpc
2018-09-18 06:01:49,934 INFO  [IPC Server handler 2 on 51943] namenode.FSNamesystem (FSNamesystem.java:checkBlocksComplete(2697)) - BLOCK* blk_1073741945_1222 is COMMITTED but not COMPLETE(numNodes= 0 <  minimum = 1) in file /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3
2018-09-18 06:01:50,191 WARN  [IPC Server handler 9 on 51943] hdfs.StateChange (FSDirAppendOp.java:appendFile(138)) - DIR* NameSystem.append: Failed to APPEND_FILE /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3 for DFSClient_NONMAPREDUCE_1981959463_7705 on 127.0.0.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_2144198447_7088 on 127.0.0.1
2018-09-18 06:01:50,191 INFO  [IPC Server handler 9 on 51943] ipc.Server (Server.java:logException(2492)) - IPC Server handler 9 on 51943, call Call#39511 Retry#0 org.apache.hadoop.hdfs.protocol.ClientProtocol.append from 127.0.0.1:49016: org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException: Failed to APPEND_FILE /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3 for DFSClient_NONMAPREDUCE_1981959463_7705 on 127.0.0.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_2144198447_7088 on 127.0.0.1
2018-09-18 06:01:50,335 INFO  [IPC Server handler 7 on 51943] hdfs.StateChange (FSDirWriteFileOp.java:completeFile(605)) - DIR* completeFile: /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2 is closed by DFSClient_NONMAPREDUCE_2144198447_7088
org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException: Failed to APPEND_FILE /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3 for DFSClient_NONMAPREDUCE_1981959463_7705 on 127.0.0.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_2144198447_7088 on 127.0.0.1
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException): Failed to APPEND_FILE /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3 for DFSClient_NONMAPREDUCE_1981959463_7705 on 127.0.0.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_2144198447_7088 on 127.0.0.1
2018-09-18 06:01:51,945 INFO  [IPC Server handler 3 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=append	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3	dst=null	perm=null	proto=rpc
2018-09-18 06:01:51,980 INFO  [IPC Server handler 4 on 51943] hdfs.StateChange (FSDirWriteFileOp.java:completeFile(605)) - DIR* completeFile: /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3 is closed by DFSClient_NONMAPREDUCE_1981959463_7705
2018-09-18 06:01:51,980 WARN  [DataStreamer for file /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3] hdfs.DataStreamer (DataStreamer.java:closeResponder(929)) - Caught exception
```

Now why I find these relevant:

1- Note that we start a block allocation with `_2`, rename it from `_2` to `_3`, and commit the block in `_3`:

```
2018-09-18 06:01:49,797 INFO  [IPC Server handler 0 on 51943] hdfs.StateChange (FSDirWriteFileOp.java:logAllocatedBlock(725)) - BLOCK* allocate blk_1073741945_1222, replicas=127.0.0.1:35859 for /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2
2018-09-18 06:01:49,853 INFO  [IPC Server handler 0 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=rename	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_2	dst=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3	perm=root:supergroup:rw-------	proto=rpc
2018-09-18 06:01:49,934 INFO  [IPC Server handler 2 on 51943] namenode.FSNamesystem (FSNamesystem.java:checkBlocksComplete(2697)) - BLOCK* blk_1073741945_1222 is COMMITTED but not COMPLETE(numNodes= 0 <  minimum = 1) in file /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3
```

This is strange as `rename` is supposed to be an atomic operation wrt the file system metadata and adding a a , but it is harmless as no data is being added.

2- The first attempt to append I see fails:

```
2018-09-18 06:01:50,191 WARN  [IPC Server handler 9 on 51943] hdfs.StateChange (FSDirAppendOp.java:appendFile(138)) - DIR* NameSystem.append: Failed to APPEND_FILE /Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3 for DFSClient_NONMAPREDUCE_1981959463_7705 on 127.0.0.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_2144198447_7088 on 127.0.0.1
```

and I assume this is the result of the rename, but I don't know that for sure.

3- There is a bit later another attempt to append that succeeds:

```
2018-09-18 06:01:51,945 INFO  [IPC Server handler 3 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true	ugi=root (auth:SIMPLE)	ip=/127.0.0.1	cmd=append	src=/Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0_3	dst=null	perm=null	proto=rpc
```

and I have not evidence that this append started before the rename, in fact, it looks like it started after the rename and targeted the new file.

--- a/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/BookKeeperIntegrationTestBase.java
+++ b/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/BookKeeperIntegrationTestBase.java
@@ -19,7 +19,6 @@ import java.nio.file.Files;
 import java.nio.file.Paths;
 import lombok.AccessLevel;
 import lombok.Getter;
-import org.apache.hadoop.service.ServiceOperations;
 
 /**
  * Base class for any StreamSegmentStore Integration Test that uses BookKeeper and RocksDB.

--- a/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/BookKeeperIntegrationTestBase.java
+++ b/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/BookKeeperIntegrationTestBase.java
@@ -11,6 +11,7 @@ package io.pravega.segmentstore.server.host;
 
 import io.pravega.common.io.FileHelpers;
 import io.pravega.segmentstore.server.store.ServiceBuilderConfig;
+import io.pravega.segmentstore.server.store.ServiceConfig;
 import io.pravega.segmentstore.server.store.StreamSegmentStoreTestBase;
 import io.pravega.segmentstore.storage.impl.rocksdb.RocksDBConfig;
 import java.io.File;
@@ -18,6 +19,7 @@ import java.nio.file.Files;
 import java.nio.file.Paths;
 import lombok.AccessLevel;
 import lombok.Getter;
+import org.apache.hadoop.service.ServiceOperations;
 
 /**
  * Base class for any StreamSegmentStore Integration Test that uses BookKeeper and RocksDB.
@@ -71,9 +73,11 @@ abstract class BookKeeperIntegrationTestBase extends StreamSegmentStoreTestBase 
      * @return A ServiceBuilderConfig instance.
      */
     protected ServiceBuilderConfig getBuilderConfig(ServiceBuilderConfig.Builder configBuilder, int instanceId) {
+        String id = Integer.toString(instanceId);
         return configBuilder
                 .makeCopy()
-                .include(RocksDBConfig.builder().with(RocksDBConfig.DATABASE_DIR, Paths.get(getRocksDBDir().toString(), Integer.toString(instanceId)).toString()))
+                .include(ServiceConfig.builder().with(ServiceConfig.INSTANCE_ID, id))
+                .include(RocksDBConfig.builder().with(RocksDBConfig.DATABASE_DIR, Paths.get(getRocksDBDir().toString(), id).toString()))
                 .build();
     }
 

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceBuilder.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceBuilder.java
@@ -11,7 +11,6 @@ package io.pravega.segmentstore.server.store;
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
-import io.pravega.segmentstore.storage.ConfigSetup;
 import io.pravega.common.concurrent.ExecutorServiceHelpers;
 import io.pravega.common.util.ConfigBuilder;
 import io.pravega.segmentstore.contracts.StreamSegmentStore;
@@ -43,6 +42,7 @@ import io.pravega.segmentstore.server.tables.TableService;
 import io.pravega.segmentstore.server.writer.StorageWriterFactory;
 import io.pravega.segmentstore.server.writer.WriterConfig;
 import io.pravega.segmentstore.storage.CacheFactory;
+import io.pravega.segmentstore.storage.ConfigSetup;
 import io.pravega.segmentstore.storage.DurableDataLogException;
 import io.pravega.segmentstore.storage.DurableDataLogFactory;
 import io.pravega.segmentstore.storage.StorageFactory;
@@ -127,11 +127,17 @@ public class ServiceBuilder implements AutoCloseable {
         this.streamSegmentStoreCreator = notConfiguredCreator(StreamSegmentStore.class);
 
         // Setup Thread Pools.
-        this.coreExecutor = executorBuilder.apply(serviceConfig.getCoreThreadPoolSize(), "core");
-        this.storageExecutor = executorBuilder.apply(serviceConfig.getStorageThreadPoolSize(), "storage-io");
+        String instancePrefix = getInstanceIdPrefix(serviceConfig);
+        this.coreExecutor = executorBuilder.apply(serviceConfig.getCoreThreadPoolSize(), instancePrefix + "core");
+        this.storageExecutor = executorBuilder.apply(serviceConfig.getStorageThreadPoolSize(), instancePrefix + "storage-io");
         this.threadPoolMetrics = new SegmentStoreMetrics.ThreadPool(this.coreExecutor);
 
         this.cacheManager = new CacheManager(serviceConfig.getCachePolicy(), this.coreExecutor);
+    }
+
+    private String getInstanceIdPrefix(ServiceConfig serviceConfig) {
+        String id = serviceConfig.getInstanceId();
+        return id == null || id.isEmpty() ? "" : id + "-";
     }
 
     //endregion

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java
@@ -18,7 +18,6 @@ import io.pravega.segmentstore.server.CachePolicy;
 import java.net.Inet4Address;
 import java.net.UnknownHostException;
 import java.time.Duration;
-import lombok.Generated;
 import lombok.Getter;
 import lombok.SneakyThrows;
 

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java
@@ -18,6 +18,7 @@ import io.pravega.segmentstore.server.CachePolicy;
 import java.net.Inet4Address;
 import java.net.UnknownHostException;
 import java.time.Duration;
+import lombok.Generated;
 import lombok.Getter;
 import lombok.SneakyThrows;
 
@@ -52,6 +53,7 @@ public class ServiceConfig {
     public static final Property<Integer> CACHE_POLICY_MAX_TIME = Property.named("cacheMaxTimeSeconds", 30 * 60);
     public static final Property<Integer> CACHE_POLICY_GENERATION_TIME = Property.named("cacheGenerationTimeSeconds", 5);
     public static final Property<Boolean> REPLY_WITH_STACK_TRACE_ON_ERROR = Property.named("replyWithStackTraceOnError", false);
+    public static final Property<String> INSTANCE_ID = Property.named("instanceId", "");
 
     public static final String COMPONENT_CODE = "pravegaservice";
 
@@ -244,6 +246,14 @@ public class ServiceConfig {
     @Getter
     private final boolean replyWithStackTraceOnError;
 
+    /**
+     * Gets a value that uniquely identifies the Service. This is useful if multiple Service Instances share the same
+     * log files or during testing, when multiple Service Instances run in the same process. This value will be prefixed
+     * to the names of all Threads used by this Service Instance, which should make log parsing easier.
+     */
+    @Getter
+    private final String instanceId;
+
     //endregion
 
     //region Constructor
@@ -299,6 +309,7 @@ public class ServiceConfig {
         int cachePolicyGenerationTime = properties.getInt(CACHE_POLICY_GENERATION_TIME);
         this.cachePolicy = new CachePolicy(cachePolicyMaxSize, Duration.ofSeconds(cachePolicyMaxTime), Duration.ofSeconds(cachePolicyGenerationTime));
         this.replyWithStackTraceOnError = properties.getBoolean(REPLY_WITH_STACK_TRACE_ON_ERROR);
+        this.instanceId = properties.get(INSTANCE_ID);
     }
 
     /**

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java
@@ -83,7 +83,7 @@ public abstract class StreamSegmentStoreTestBase extends ThreadPooledTestSuite {
     private static final int TRANSACTIONS_PER_SEGMENT = 1;
     private static final int APPENDS_PER_SEGMENT = 100;
     private static final int ATTRIBUTE_UPDATES_PER_SEGMENT = 100;
-    private static final int MAX_INSTANCE_COUNT = 5;
+    private static final int MAX_INSTANCE_COUNT = 4;
     private static final List<UUID> ATTRIBUTES = Arrays.asList(Attributes.EVENT_COUNT, UUID.randomUUID(), UUID.randomUUID());
     private static final int EXPECTED_ATTRIBUTE_VALUE = APPENDS_PER_SEGMENT + ATTRIBUTE_UPDATES_PER_SEGMENT;
     private static final Duration TIMEOUT = Duration.ofSeconds(120);

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java
@@ -9,7 +9,6 @@
  */
 package io.pravega.segmentstore.server.store;
 
-import com.google.common.collect.Iterators;
 import io.pravega.common.Exceptions;
 import io.pravega.common.ObjectClosedException;
 import io.pravega.common.TimeoutTimer;

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java
@@ -261,9 +261,6 @@ public abstract class StreamSegmentStoreTestBase extends ThreadPooledTestSuite {
             val segmentNames = createSegments(context.getActiveStore());
             val segmentsAndTransactions = new ArrayList<String>(segmentNames);
             log.info("Created Segments: {}.", String.join(", ", segmentNames));
-            val transactionsBySegment = createTransactions(segmentNames, context.getActiveStore());
-            transactionsBySegment.values().forEach(segmentsAndTransactions::addAll);
-            log.info("Created Transactions: {}.", transactionsBySegment.values().stream().flatMap(Collection::stream).collect(Collectors.joining(", ")));
 
             // Generate all the requests.
             HashMap<String, Long> lengths = new HashMap<>();
@@ -271,12 +268,10 @@ public abstract class StreamSegmentStoreTestBase extends ThreadPooledTestSuite {
             HashMap<String, ByteArrayOutputStream> segmentContents = new HashMap<>();
             val appends = createAppendDataRequests(segmentsAndTransactions, segmentContents, lengths,
                     applyFencingMultiplier(ATTRIBUTE_UPDATES_PER_SEGMENT), applyFencingMultiplier(APPENDS_PER_SEGMENT));
-            val mergers = createMergeTransactionsRequests(transactionsBySegment, lengths, segmentContents);
-            val seals = createSealSegmentsRequests(segmentNames);
-            val requests = Iterators.concat(appends.iterator(), mergers.iterator(), seals.iterator());
+            val requests = appends.iterator();
 
             // Calculate how frequently to create a new instance of the Segment Store.
-            int newInstanceFrequency = (appends.size() + mergers.size() + seals.size()) / applyFencingMultiplier(MAX_INSTANCE_COUNT);
+            int newInstanceFrequency = appends.size() / applyFencingMultiplier(MAX_INSTANCE_COUNT);
             log.info("Creating a new Segment Store instance every {} operations.", newInstanceFrequency);
 
             // Execute all the requests.

--- a/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java
+++ b/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java
@@ -148,8 +148,6 @@ public class RollingStorage implements SyncStorage {
             log.debug("Handle refreshed: {}.", h);
         }
 
-        Preconditions.checkArgument(offset < h.length(), "Offset %s is beyond the last offset %s of the segment.",
-                offset, h.length());
         Preconditions.checkArgument(offset + length <= h.length(), "Offset %s + length %s is beyond the last offset %s of the segment.",
                 offset, length, h.length());
 
@@ -510,7 +508,22 @@ public class RollingStorage implements SyncStorage {
         Preconditions.checkArgument(!handle.isSealed(), "Cannot rollover a Sealed Segment.");
         log.debug("Rolling over '{}'.", handle);
         sealActiveChunk(handle);
-        createChunk(handle);
+        try {
+            createChunk(handle);
+        } catch (StreamSegmentExistsException ex) {
+            // It may be possible that a concurrent rollover request using a different handle (either from this instance
+            // or another) has already created the new chunk. Refresh the handle and try again. This is usually the case
+            // with concurrent Storage instances trying to modify the same segment at the same time.
+            int chunkCount = handle.chunks().size();
+            handle.refresh(openHandle(handle.getSegmentName(), false));
+            if (chunkCount == handle.chunks().size()) {
+                // Nothing changed; re-throw the exception.
+                throw ex;
+            } else {
+                // We've just refreshed the handle and picked up the latest chunk. Move on.
+                log.warn("Aborted rollover due to concurrent rollover detected ('{}').", handle);
+            }
+        }
     }
 
     private void sealActiveChunk(RollingSegmentHandle handle) throws StreamSegmentException {

